src/header/common/content_disposition.rs

```diff
@@ -65,88 +218,215 @@ pub enum DispositionParam {
-///       None,
-///       "\u{2764}".as_bytes().to_vec()
-///     )]
+///     disposition: DispositionType::FormData,
+///     parameters: vec![
+///         DispositionParam::Name(String::from("file")),
+///         DispositionParam::Filename(String::from("bill.odt")),
+///     ],
 /// };
+/// assert_eq!(cd2.get_name(), Some("file")); // field name
+/// assert_eq!(cd2.get_filename(), Some("bill.odt"));
 /// ```
+///
+/// # WARN
+/// If "filename" parameter is supplied, do not use the file name blindly, check and possibly
+/// change to match local file system conventions if applicable, and do not use directory path
+/// information that may be present. See [RFC2183](https://tools.ietf.org/html/rfc2183#section-2.3)
+/// .
 #[derive(Clone, Debug, PartialEq)]
 pub struct ContentDisposition {
-    /// The disposition
+    /// The disposition type
     pub disposition: DispositionType,
     /// Disposition parameters
     pub parameters: Vec<DispositionParam>,
 }
+
 impl ContentDisposition {
-    /// Parse a raw Content-Disposition header value
+    /// Parse a raw Content-Disposition header value.
     pub fn from_raw(hv: &header::HeaderValue) -> Result<Self, ::error::ParseError> {
-        header::from_one_raw_str(Some(hv)).and_then(|s: String| {
-            let mut sections = s.split(';');
-            let disposition = match sections.next() {
-                Some(s) => s.trim(),
-                None => return Err(::error::ParseError::Header),
-            };
-
-            let mut cd = ContentDisposition {
-                disposition: if disposition.eq_ignore_ascii_case("inline") {
-                    DispositionType::Inline
-                } else if disposition.eq_ignore_ascii_case("attachment") {
-                    DispositionType::Attachment
-                } else {
-                    DispositionType::Ext(disposition.to_owned())
-                },
-                parameters: Vec::new(),
-            };
-
-            for section in sections {
-                let mut parts = section.splitn(2, '=');
+        // `header::from_one_raw_str` invokes `hv.to_str` which assumes `hv` contains only visible
+        //  ASCII characters. So `hv.as_bytes` is necessary here.
+        let hv = String::from_utf8(hv.as_bytes().to_vec())
+            .map_err(|_| ::error::ParseError::Header)?;
+        let (disp_type, mut left) = split_once_and_trim(hv.as_str().trim(), ';');
+        if disp_type.len() == 0 {
+            return Err(::error::ParseError::Header);
+        }
+        let mut cd = ContentDisposition {
+            disposition: disp_type.into(),
+            parameters: Vec::new(),
+        };
 
-                let key = if let Some(key) = parts.next() {
-                    key.trim()
+        while left.len() > 0 {
+            let (param_name, new_left) = split_once_and_trim(left, '=');
+            if param_name.len() == 0 {
+                return Err(::error::ParseError::Header);
+            }
```

Comment:
Should we check for `new_left != 0`?

Reply:
Yes but not necessary. Line 294 `header::parse_extended_value` will error out if `ext_value` is empty, which includes cases where `new_left.len() == 0`. The nested if block starting at L305 falls into the branch starting at L334, if `new_left` is empty, which accepts empty values. But check in L342 will ensure it error out.

Reply:
But if `new_left.len() == 0` it wouldn't fall into the line 289 branch.
I just feel like it would be better to leave earlier in bad case instead of checking it out.
Though I'm not sure should we just ignore such parameter or error....

Reply:
Why wouldn't? Whether falling into L289 only dependent on the presence of trailing asterisk of `param_name`. 

Fine, I will add that check in a new commit.

I am also not sure and fail to find related statements in those standards or elsewhere. But I do not think it is really a problem in practice because that is generated by browsers and does not rely on end-users input. Besides, what the lenient parsing implementation wants to do is just ensuring whatever directly submitted by end-users (typically, file names) are parsed correctly as long as browsers produce headers correctly.

Reply:
If you think that the current version is ok as it is unlikely to happen then leave it as it is 😄